### PR TITLE
Tilee/fix sdl injection bug

### PR DESCRIPTION
**Pull request by MS-TimothyMothra**
Fix for issue discovered in SDL review. Will guard against arbitrary length strings received from outside the application.

- [x] I ran Unit Tests locally.

### CODE REVIEW

File: Src/Common/StringUtilities.cs [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+    {
+        /// <summary>
+        /// Check a strings length and trim to a max length if needed.
+        /// </summary>
+        public static string EnforceMaxLength(string input, int maxLength)
+        {
+            Debug.Assert(input != null, $"{nameof(input)} must not be null");
+            Debug.Assert(maxLength > 0, $"{nameof(maxLength)} must be greater than 0");
+
+            if (input != null && input.Length > maxLength)
```
[the remark this answers is not in the document]
Comment (MS-TimothyMothra) [reply Src/Common/StringUtilities.cs]: I saw this design pattern used in other places.
Not sure about best practices in this case.

File: Src/Common/CorrelationIdLookupHelper.cs [hunk 2]
```diff
@@ -190,8 +190,12 @@ public bool IsFetchAppInProgress(string ikey)
         /// </summary>
         /// <param name="ikey">Instrumentation Key is expected to be a Guid string.</param>
         /// <param name="appId">Application Id is expected to be a Guid string. App Id needs to be Http Header safe, and all non-ASCII characters will be removed.</param>
+        /// <remarks>To protect against injection attacks, AppId will be truncated to a maximum length.</remarks>
         private void GenerateCorrelationIdAndAddToDictionary(string ikey, string appId)
         {
+            // Arbitrary maximum length to guard against injections.
+            appId = StringUtilities.EnforceMaxLength(appId, InjectionGuardConstants.AppIdMaxLengeth);
```
[the remark this answers is not in the document]
Comment (MS-TimothyMothra) [reply Src/Common/CorrelationIdLookupHelper.cs]: I don't understand your "less resources" comment.

If we drop the AppId, the sdk will keep making appId requests (after brief timeout). I think either approach could be valid, but lets discuss what's best for this use case.

[the remark this answers is not in the document]
Comment (Dmitry-Matveev) [reply Src/Common/CorrelationIdLookupHelper.cs]: There are two ways we can go about appId if we think it's wrong (e.g. does not satisfy length or format):
- Drop it and let it query for it again later;
- Accept it (e.g. cut to length or ignore the wrong format) and let SDK process the wrong appID entry everywhere down the line.
Both cases have their own pros, so I was interested in which one is "cheaper" from CPU/Memory perspective - trying out for new appId or keep re-using the wrong one on related code paths.

[the remark this answers is not in the document]
Comment (MS-TimothyMothra) [reply Src/Common/CorrelationIdLookupHelper.cs]: Discussed offline with Dmitry.
We are deciding to truncate the value and accept it as is.
Assuming that if an attacker injected a fraudulent value, that attacker would continue to inject fraudulent values on retry attempts. Best to quit and wait until app pool recycles to try again.

File: Src/Common/StringUtilities.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (MS-TimothyMothra) [reply Src/Common/StringUtilities.cs]: Discussed with Dmitry offline.
We're going to leave this for now, both conditions are already addressed in the if statement so no risk to the application.
At a later date, we'll add a Test Project around our Common library and remove these asserts.

File: Src/Common/InjectionGuardConstants.cs [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+    {
+        /// <summary>
+        /// Max length of AppId allowed in response from Breeze.
+        /// </summary>
+        public const int AppIdMaxLengeth = 50;
+
+        /// <summary>
+        /// Max length of incoming Request Header value allowed.
+        /// </summary>
+        public const int RequestHeaderMaxLength = 100;
```
[the remark this answers is not in the document]
Comment (SergeyKanzhelev) [reply Src/Common/InjectionGuardConstants.cs]: Today it only has `app-id=v1_guid` or something like this. So `100` may be enough. I believe @iusafaro removed role name reading/setting logic already so no role name in this header any longer.

We will definitely be in trouble when we will want to send more than a single property. So maybe allow up to `1Kb` to be safe, but not break potential scenarios.

Note, this header will be gone with the new W3C headers.

**Pull request merged**